File: modules/core/src/events/event_manager.js
import {isBlank, BaseException, isPresent} from 'facade/lang';
import {Element} from 'facade/dom';
import {List, ListWrapper, MapWrapper} from 'facade/collection';
import {VmTurnZone} from 'core/zone/vm_turn_zone';

export class EventManager {
  plugins: List<EventManagerPlugin>;
  zone: VmTurnZone;

  constructor(plugins: List<EventManagerPlugin>, zone: VmTurnZone) {
    this.zone = zone;
    this.plugins = ListWrapper.create();

    for (var i = 0; i < plugins.length; i++) {
      var plugin = ListWrapper.get(plugins, i);
      if (plugin.isEnabled()) {
        plugin.manager = this;
        ListWrapper.push(this.plugins, plugin);
      }
    }
  }

  addEventListener(element: Element, eventName: string, handler: Function) {
    var plugin = this._findPluginFor(eventName);

    if (isBlank(plugin)) {
      throw new BaseException(`no plugin found to handle the ${eventName} event`);
    }

    plugin.addEventListener(element, eventName, handler);
  }

  _findPluginFor(eventName: string): EventManagerPlugin {
    var plugins = this.plugins;
    for (var i = 0; i < plugins.length; i++) {
      var plugin = plugins[i];
      if (plugin.supports(eventName)) {
        return plugin;
      }
    }
    return null;
  }
}

export class EventManagerPlugin {
  manager: EventManager;

  isEnabled(): boolean {
    return true;
  }

  supports(eventName: string): boolean {
    return false;
  }

  addEventListener(element: Element, eventName: string, handler: Function) {
    throw "not implemented";
  }
}

File: lib/src/platform/dom/events/event_manager.dart
library angular2.src.platform.dom.events.event_manager;

import "package:angular2/src/facade/exceptions.dart"
    show BaseException, WrappedException;
import "package:angular2/src/core/di.dart" show Injectable, Inject, OpaqueToken;
import "package:angular2/src/core/zone/ng_zone.dart" show NgZone;
import "package:angular2/src/facade/collection.dart" show ListWrapper;

const OpaqueToken EVENT_MANAGER_PLUGINS =
    const OpaqueToken("EventManagerPlugins");

@Injectable()
class EventManager {
  NgZone _zone;
  List<EventManagerPlugin> _plugins;
  EventManager(@Inject(EVENT_MANAGER_PLUGINS) List<EventManagerPlugin> plugins,
      this._zone) {
    plugins.forEach((p) => p.manager = this);
    this._plugins = ListWrapper.reversed(plugins);
  }
  addEventListener(dynamic element, String eventName, Function handler) {
    var plugin = this._findPluginFor(eventName);
    plugin.addEventListener(element, eventName, handler);
  }

  Function addGlobalEventListener(
      String target, String eventName, Function handler) {
    var plugin = this._findPluginFor(eventName);
    return plugin.addGlobalEventListener(target, eventName, handler);
  }

  NgZone getZone() {
    return this._zone;
  }

  /** @internal */
  EventManagerPlugin _findPluginFor(String eventName) {
    var plugins = this._plugins;
    for (var i = 0; i < plugins.length; i++) {
      var plugin = plugins[i];
      if (plugin.supports(eventName)) {
        return plugin;
      }
    }
    throw new BaseException(
        '''No event manager plugin found for event ${ eventName}''');
  }
}

class EventManagerPlugin {
  EventManager manager;
  // That is equivalent to having supporting $event.target
  bool supports(String eventName) {
    return false;
  }

  addEventListener(dynamic element, String eventName, Function handler) {
    throw "not implemented";
  }

  Function addGlobalEventListener(
      String element, String eventName, Function handler) {
    throw "not implemented";
  }
}
